File: chaocipher.go
package main

import (
	"flag"
	"strings"
	"fmt"
	"os"
	"io/ioutil"
	"unicode/utf8"
)

var (
	decrypt = flag.Bool("d", false, "Decrypt instead of Encrypt.")
)


func main() {
	flag.Parse()

	data, _ := ioutil.ReadAll(os.Stdin)
	b := strings.TrimSuffix(string(data), "\r\n")
	b = strings.TrimSuffix(b, "\n")
	b = strings.ToUpper(b)
	if *decrypt {
		plainText := Chao(string(b), Decrypt, true)
		fmt.Println(plainText)
	} else {
		cipherText := Chao(string(b), Encrypt, true)
		fmt.Println(cipherText)
	} 
}

type Mode int

const (
	Encrypt Mode = iota
	Decrypt
)

const (
	lAlphabet = "HXUCZVAMDSLKPEFJRIGTWOBNYQ"
	rAlphabet = "PTLNBQDEOYSFAVZKGJRIHWXUMC"
)

func Chao(text string, mode Mode, showSteps bool) string {
	len := len(text)
	if utf8.RuneCountInString(text) != len {
		fmt.Println("Text contains non-ASCII characters")
		return ""
	}
	left := lAlphabet
	right := rAlphabet
	eText := make([]byte, len)
	temp := make([]byte, 26)

	for i := 0; i < len; i++ {
		if showSteps {
			fmt.Fprintln(os.Stderr, left, " ", right)
		}
		var index int
		if mode == Encrypt {
			index = strings.IndexByte(right, text[i])
			eText[i] = left[index]
		} else {
			index = strings.IndexByte(left, text[i])
			eText[i] = right[index]
		}
		if i == len-1 {
			break
		}

		for j := index; j < 26; j++ {
			temp[j-index] = left[j]
		}
		for j := 0; j < index; j++ {
			temp[26-index+j] = left[j]
		}
		store := temp[1]
		for j := 2; j < 14; j++ {
			temp[j-1] = temp[j]
		}
		temp[13] = store
		left = string(temp[:])

		for j := index; j < 26; j++ {
			temp[j-index] = right[j]
		}
		for j := 0; j < index; j++ {
			temp[26-index+j] = right[j]
		}
		store = temp[0]
		for j := 1; j < 26; j++ {
			temp[j-1] = temp[j]
		}
		temp[25] = store
		store = temp[2]
		for j := 3; j < 14; j++ {
			temp[j-1] = temp[j]
		}
		temp[13] = store
		right = string(temp[:])
	}
	return string(eText[:])
}

File: chaocipher.php
<?php

function chao($text, $mode = 'encrypt', $showSteps = true)
{
    $text = strtoupper($text);
    if (preg_match('/[^A-Z]/', $text)) {
        fwrite(STDERR, "Text contains non-ASCII uppercase letters only\n");
        return '';
    }

    $left = str_split("HXUCZVAMDSLKPEFJRIGTWOBNYQ");
    $right = str_split("PTLNBQDEOYSFAVZKGJRIHWXUMC");
    $output = [];

    for ($i = 0; $i < strlen($text); $i++) {
        if ($showSteps) {
            fwrite(STDERR, implode('', $left) . "  " . implode('', $right) . PHP_EOL);
        }

        $ch = $text[$i];
        if ($mode === 'encrypt') {
            $index = array_search($ch, $right);
            $output[] = $left[$index];
        } else {
            $index = array_search($ch, $left);
            $output[] = $right[$index];
        }

        if ($i === strlen($text) - 1) {
            break;
        }

        // Permutação esquerda
        $left = rotateArray($left, $index);
        $store = $left[1];
        for ($j = 2; $j < 14; $j++) {
            $left[$j - 1] = $left[$j];
        }
        $left[13] = $store;

        // Permutação direita
        $right = rotateArray($right, $index);
        $store = $right[0];
        for ($j = 1; $j < 26; $j++) {
            $right[$j - 1] = $right[$j];
        }
        $right[25] = $store;
        $store = $right[2];
        for ($j = 3; $j < 14; $j++) {
            $right[$j - 1] = $right[$j];
        }
        $right[13] = $store;
    }

    return implode('', $output);
}

function rotateArray($arr, $index)
{
    $len = count($arr);
    return array_merge(array_slice($arr, $index), array_slice($arr, 0, $index));
}

// ===== CLI entry point =====

$options = getopt("d");
$decrypt = isset($options['d']);

$stdin = '';
while (!feof(STDIN)) {
    $stdin .= fread(STDIN, 1024);
}
$stdin = strtoupper(trim($stdin));

if ($decrypt) {
    echo chao($stdin, 'decrypt', true) . PHP_EOL;
} else {
    echo chao($stdin, 'encrypt', true) . PHP_EOL;
}
